File: Assignments/02_Heuristic_Solutions/02_Path_Length.py
def path_length(src, dst):
    if src == dst:
        return 0
    for edge in Edge:
        if edge[0] == src and edge[1] == dst:
            return edge[2]

    for edge in Edge:
        if edge[0] == src:
            adj_node = edge[1]
            if adj_node not in visited:
                visited.append(adj_node)
                cost = path_length(adj_node, dst)
                if cost != -1:
                    return edge[2] + cost
    return -1


# Main

# Graph
Edge = [
    ("i", "a", 35),
    ("i", "b", 45),
    ("a", "c", 22),
    ("a", "d", 32),
    ("b", "d", 28),
    ("b", "e", 36),
    ("b", "f", 27),
    ("c", "d", 31),
    ("c", "g", 47),
    ("d", "g", 30),
    ("e", "g", 26),
]

visited = []

src, dst = input("Enter Source and Destination: ").split()
print("Cost = ", path_length(src, dst))

File: Assignments/02_Heuristic_Solutions/02_Path_Length.pl
:- write('The program has been consulted').

% KB (Graph)
edge(i, a, 35).
edge(i, b, 45).
edge(a, c, 22).
edge(a, d, 32).
edge(b, d, 28).
edge(b, e, 36).
edge(b, f, 27).
edge(c, d, 31).
edge(c, g, 47).
edge(d, g, 30).
edge(e, g, 26).

% Rules and Predicates

length(Src, Dst, Cost):- edge(Src, Dst, Cost),!.
length(Src, Dst, Cost):- edge(Src, Node, CostSrcNode), length(Node, Dst, CostDstNode), Cost is (CostSrcNode+ CostDstNode).
